examples/mlperf/rnnt/train.py

```diff
@@ -0,0 +1,367 @@
+# %%
+from tinygrad.tensor import Tensor
+from temp.print_tree import print_tree
+from tinygrad.nn.optim import Adam, LAMB
+
+# %%
+import functools
+import itertools
+
+# %%
+from tinygrad.helpers import dtypes
+from tinygrad.ops import Device
+
+# Device.DEFAULT = 'METAL'
+
+# %% [markdown]
+# # data
+
+# %%
+import pathlib
+import json
+import  numpy as np
+import matplotlib.pyplot as plt
+import librosa
+import soundfile
+BASEDIR = pathlib.Path("../../../extra/datasets/librispeech")
+with open(BASEDIR / "dev-clean-wav.json") as f:
+  ci = json.load(f)
+FILTER_BANK = np.expand_dims(librosa.filters.mel(sr=16000, n_fft=512, n_mels=80, fmin=0, fmax=8000), 0)
+WINDOW = librosa.filters.get_window("hann", 320)
+
+def feature_extract(x, x_lens):
+  x_lens = np.ceil((x_lens / 160) / 3).astype(np.int32)
+
+  # pre-emphasis
+  x = np.concatenate((np.expand_dims(x[:, 0], 1), x[:, 1:] - 0.97 * x[:, :-1]), axis=1)
+
+  # stft
+  x = librosa.stft(x, n_fft=512, window=WINDOW, hop_length=160, win_length=320, center=True, pad_mode="reflect")
+  x = np.stack((x.real, x.imag), axis=-1)
+
+  # power spectrum
+  x = (x**2).sum(-1)
+
+  # mel filter bank
+  x = np.matmul(FILTER_BANK, x)
+
+  # log
+  x = np.log(x + 1e-20)
+
+  # feature splice
+  seq = [x]
+  for i in range(1, 3):
+    tmp = np.zeros_like(x)
+    tmp[:, :, :-i] = x[:, :, i:]
+    seq.append(tmp)
+  features = np.concatenate(seq, axis=1)[:, :, ::3]
+
+  # normalize
+  features_mean = np.zeros((features.shape[0], features.shape[1]), dtype=np.float32)
+  features_std = np.zeros((features.shape[0], features.shape[1]), dtype=np.float32)
+  for i in range(features.shape[0]):
+    features_mean[i, :] = features[i, :, :x_lens[i]].mean(axis=1)
+    features_std[i, :] = features[i, :, :x_lens[i]].std(axis=1, ddof=1)
+  features_std += 1e-5
+  features = (features - np.expand_dims(features_mean, 2)) / np.expand_dims(features_std, 2)
```

Comment:
So all the numpy operations here could be very easily transferred to tinygrad Tensors except the `librosa.stft` part. Especially if they are being called periodically, you would like to eventually JIT this part of the processing to speed it up.

Reply:
I just copied this code from https://github.com/tinygrad/tinygrad/blob/master/extra/datasets/librispeech.py can translate to tensors but this is all just a couple of seconds for the whole dataset.